stl/src/tzdb.cpp

```diff
@@ -1,15 +1,345 @@
+
+    _Info->_Links = new (_STD nothrow) const char*[_Info->_Num_time_zones];
+    if (_Info->_Links == nullptr) {
+        return nullptr;
+    }
+
+    // init to ensure __std_tzdb_delete_time_zones() cleanup is valid
+    _STD fill_n(_Info->_Links, static_cast<ptrdiff_t>(_Info->_Num_time_zones), nullptr);
+
+    for (size_t _Name_idx = 0; _Name_idx < _Info->_Num_time_zones; ++_Name_idx) {
+        int32_t _Elem_len{};
+        const auto* _Elem = __icu_uenum_unext(_Enum.get(), &_Elem_len, &_UErr);
+        if (U_FAILURE(_UErr) || _Elem == nullptr) {
+            return _Report_error(_Info, __std_tzdb_error::_Icu_error);
+        }
+
+        _Info->_Names[_Name_idx] = _Allocate_wide_to_narrow(_Elem, _Elem_len, _Info->_Err);
+        if (_Info->_Names[_Name_idx] == nullptr) {
+            return _Propagate_error(_Info);
+        }
+
+        int32_t _Link_len{};
+        const auto _Link = _Get_canonical_id(_Elem, _Elem_len, _Link_len, _Info->_Err);
+        if (_Link == nullptr) {
+            return _Propagate_error(_Info);
+        }
+
+        if (_STD u16string_view{_Elem, static_cast<size_t>(_Elem_len)}
+            != _STD u16string_view{_Link.get(), static_cast<size_t>(_Link_len)}) {
+            _Info->_Links[_Name_idx] = _Allocate_wide_to_narrow(_Link.get(), _Link_len, _Info->_Err);
+            if (_Info->_Links[_Name_idx] == nullptr) {
+                return _Propagate_error(_Info);
+            }
+        }
+    }
+
+    return _Info.release();
+}
+
+void __stdcall __std_tzdb_delete_time_zones(__std_tzdb_time_zones_info* _Info) noexcept {
+    if (_Info != nullptr) {
+        if (_Info->_Names != nullptr) {
+            for (size_t _Idx = 0; _Idx < _Info->_Num_time_zones; ++_Idx) {
+                delete[] _Info->_Names[_Idx];
+            }
+
+            delete[] _Info->_Names;
+            _Info->_Names = nullptr;
+        }
+
+        if (_Info->_Links != nullptr) {
+            for (size_t _Idx = 0; _Idx < _Info->_Num_time_zones; ++_Idx) {
+                delete[] _Info->_Links[_Idx];
+            }
+
+            delete[] _Info->_Names;
+            _Info->_Names = nullptr;
+        }
+    }
+}
+
+_NODISCARD __std_tzdb_current_zone_info* __stdcall __std_tzdb_get_current_zone() noexcept {
+    // On exit---
+    //    _Info == nullptr          --> bad_alloc()
+    //    _Info->_Err == _Win_error --> failed, call GetLastError()
+    //    _Info->_Err == _Icu_error --> runtime_error interacting with ICU
+    _STD unique_ptr<__std_tzdb_current_zone_info, decltype(&__std_tzdb_delete_current_zone)> _Info{
+        new (_STD nothrow) __std_tzdb_current_zone_info{}, &__std_tzdb_delete_current_zone};
+    if (_Info == nullptr) {
+        return nullptr;
+    }
+
+    if (_Acquire_icu_functions() < _Icu_api_level::_Has_icu_addresses) {
+        return _Report_error(_Info, __std_tzdb_error::_Win_error);
+    }
+
+    UErrorCode _Err{};
+    char16_t _Id_buf[256];

```

Comment:
You mentioned that this is an arbitrary size, and this code can be changed to follow the other pattern of "query the size, then allocate".